File: pitnn_inference.cpp
/*
 * PITNN DAB Converter — C++ Inference (Option 2)
 * ================================================
 * Real-time PITNN controller using LibTorch.
 * Runs on NVIDIA Jetson, embedded Linux, or any platform
 * with the LibTorch C++ library installed.
 *
 * Required files (same directory as executable):
 *     pitnn_scripted.pt   — TorchScript model
 *     pitnn_mu.npy        — normalisation means    (load via load_npy_array)
 *     pitnn_sigma.npy     — normalisation std devs (load via load_npy_array)
 *
 * Build:
 *     mkdir build && cd build
 *     cmake -DCMAKE_PREFIX_PATH=/path/to/libtorch ..
 *     make -j4
 *
 * Run:
 *     ./pitnn_inference
 *
 * Copyright (c) 2026 Chukwuemeka Nzeadibe
 * Mississippi State University — All Rights Reserved
 */

#include <torch/script.h>
#include <iostream>
#include <array>
#include <vector>
#include <string>
#include <fstream>
#include <cmath>
#include <chrono>
#include <cassert>

// ─────────────────────────────────────────────────────────────────────────────
// CONSTANTS  (must match values used during training)
// ─────────────────────────────────────────────────────────────────────────────
static constexpr float V1_NOM    = 800.0f;
static constexpr float V2_NOM    = 800.0f;
static constexpr float FSW       = 100000.0f;
static constexpr float PI_F      = 3.14159265358979f;
static constexpr float PHI12_MIN = PI_F * 0.65f;   // 2.0420 rad — lower bound phi1/phi2
static constexpr float PHI12_MAX = PI_F * 0.99f;   // 3.1102 rad — upper bound phi1/phi2
static constexpr float PHI12_NOM = PI_F * 0.95f;   // nominal seed for buffer priming
static constexpr float PHI_MIN   = 0.02f;           // lower bound for phi3
static constexpr float PHI3_MAX  = 1.50f;
static constexpr int   SEQ_LEN   = 20;
static constexpr int   N_FEAT    = 8;

// ─────────────────────────────────────────────────────────────────────────────
// NORMALISATION CONSTANTS
// NOTE: phi1/phi2 now vary across [PHI12_MIN, PHI12_MAX] — the mu/sigma values
// below MUST be regenerated after retraining. Run:
//     python pitnn_inspect_exports.py
// and paste the printed C++ arrays here.
// Feature order: [V1, V2, iL, phi1, phi2, phi3, Pref, V1V2/Vnom2]
// ─────────────────────────────────────────────────────────────────────────────
static const float MU[N_FEAT] = {
    800.02514648f,   // V1   (V)
    800.20202637f,   // V2   (V)
     25.38187981f,   // iL   (A)
      2.99313903f,   // phi1 (rad)
      2.99313903f,   // phi2 (rad)
      0.47853500f,   // phi3 (rad)
  37593.66796875f,   // Pref (W)
      1.00037396f    // V1*V2/Vnom2
};

static const float SIGMA[N_FEAT] = {
     46.25230789f,   // V1
     46.19021606f,   // V2
     15.18008804f,   // iL
      0.00862500f,   // phi1
      0.00862500f,   // phi2
      0.29205400f,   // phi3
  18769.15234375f,   // Pref
      0.08148100f    // V1*V2/Vnom2
};


// ─────────────────────────────────────────────────────────────────────────────
// SIMPLE .npy READER  (for float32 1D arrays only)
// ─────────────────────────────────────────────────────────────────────────────
static bool load_npy_float32(const std::string& path,
                              std::vector<float>& out) {
    std::ifstream f(path, std::ios::binary);
    if (!f) { std::cerr << "Cannot open: " << path << std::endl; return false; }

    // Skip npy magic (6) + version (2) + header_len (2) = 10 bytes, then header
    char magic[7]; f.read(magic, 6);
    uint8_t major, minor; f.read((char*)&major, 1); f.read((char*)&minor, 1);
    uint16_t hlen; f.read((char*)&hlen, 2);
    std::string header(hlen, ' '); f.read(&header[0], hlen);

    // Find the shape from the header string
    size_t pos = header.find("'shape': (");
    if (pos == std::string::npos) { std::cerr << "Bad npy header\n"; return false; }
    size_t end = header.find(")", pos);
    std::string shape_str = header.substr(pos + 10, end - pos - 10);
    // Remove trailing comma if present
    if (!shape_str.empty() && shape_str.back() == ',')
        shape_str.pop_back();
    size_t n = std::stoul(shape_str);

    out.resize(n);
    f.read((char*)out.data(), n * sizeof(float));
    return true;
}


// ─────────────────────────────────────────────────────────────────────────────
// PITNN CONTROLLER CLASS
// ─────────────────────────────────────────────────────────────────────────────
class PITNNController {
public:
    // Constructor — loads model from disk once
    explicit PITNNController(const std::string& model_path = "pitnn_scripted.pt",
                              bool use_cuda = false) {
        _device = (use_cuda && torch::cuda::is_available())
                  ? torch::kCUDA : torch::kCPU;

        std::cout << "[PITNN] Loading: " << model_path << std::endl;
        try {
            _model = torch::jit::load(model_path, _device);
            _model.eval();
        } catch (const c10::Error& e) {
            throw std::runtime_error(
                std::string("Failed to load model: ") + e.what());
        }

        // Copy compile-time constants into instance arrays
        for (int i = 0; i < N_FEAT; i++) {
            _mu[i]    = MU[i];
            _sigma[i] = SIGMA[i];
        }

        reset();
        std::cout << "[PITNN] Ready on "
                  << (_device == torch::kCUDA ? "CUDA" : "CPU") << std::endl;
    }

    // Clear history buffer — call when starting a new operating scenario
    void reset() {
        for (int t = 0; t < SEQ_LEN; t++)
            for (int f = 0; f < N_FEAT; f++)
                _buffer[t][f] = 0.0f;
        _phi1_prev = PHI12_NOM;   // seed with nominal inner duty
        _phi2_prev = PHI12_NOM;
        _phi3_prev = 0.22f;
    }

    /*
     * step() — call once per switching cycle
     *
     * Parameters:
     *   V1    : primary DC bus voltage (V)   — from ADC
     *   V2    : secondary DC bus voltage (V) — from ADC
     *   iL    : inductor current (A)         — from current sensor
     *   Pref  : power reference (W)          — from outer PI loop
     *
     * Returns:
     *   {phi1, phi2, phi3} in radians — all three independently predicted
     *   phi1 ∈ [PHI12_MIN, PHI12_MAX] — primary bridge inner duty
     *   phi2 ∈ [PHI12_MIN, PHI12_MAX] — secondary bridge inner duty
     *   phi3 ∈ [PHI_MIN,   PHI3_MAX]  — external phase shift → gate drive delay
     */
    std::array<float, 3> step(float V1, float V2, float iL, float Pref) {

        // Build 8-feature state vector using previous predicted angles
        float v_ratio = (V1 * V2) / (V1_NOM * V2_NOM);
        float feat[N_FEAT] = {
            V1, V2, iL, _phi1_prev, _phi2_prev, _phi3_prev, Pref, v_ratio
        };

        // Normalise: (x - mu) / sigma
        float feat_norm[N_FEAT];
        for (int i = 0; i < N_FEAT; i++)
            feat_norm[i] = (feat[i] - _mu[i]) / _sigma[i];

        // Shift buffer left by one step, append new state at end
        for (int t = 0; t < SEQ_LEN - 1; t++)
            for (int f = 0; f < N_FEAT; f++)
                _buffer[t][f] = _buffer[t+1][f];
        for (int f = 0; f < N_FEAT; f++)
            _buffer[SEQ_LEN-1][f] = feat_norm[f];

        // Build input tensor (1, SEQ_LEN, N_FEAT) — copy from buffer
        auto opts  = torch::TensorOptions().dtype(torch::kFloat32);
        auto input = torch::from_blob(_buffer,
                                       {1, SEQ_LEN, N_FEAT}, opts)
                         .clone()
                         .to(_device);

        // PITNN forward pass
        torch::NoGradGuard no_grad;
        auto output = _model.forward({input}).toTensor()
                            .squeeze()
                            .to(torch::kCPU);

        float phi1 = output[0].item<float>();
        float phi2 = output[1].item<float>();
        float phi3 = output[2].item<float>();

        _phi1_prev = phi1;
        _phi2_prev = phi2;
        _phi3_prev = phi3;
        return {phi1, phi2, phi3};
    }

    // Convert phi3 (rad) to gate drive phase delay in microseconds
    static float phi3_to_delay_us(float phi3) {
        return phi3 / (2.0f * PI_F * FSW) * 1e6f;
    }

    // Convert phi1 (rad) to inner duty cycle percentage
    static float phi1_to_duty_pct(float phi1) {
        return (phi1 / PI_F) * 100.0f;
    }

private:
    torch::jit::script::Module _model;
    torch::Device _device{torch::kCPU};
    float _mu[N_FEAT];
    float _sigma[N_FEAT];
    float _buffer[SEQ_LEN][N_FEAT];
    float _phi1_prev;
    float _phi2_prev;
    float _phi3_prev;
};


// ─────────────────────────────────────────────────────────────────────────────
// DEMO
// ─────────────────────────────────────────────────────────────────────────────
int main(int argc, char* argv[]) {
    std::cout << "====================================================" << std::endl;
    std::cout << "  PITNN DAB Converter — C++ Inference (Option 2)"     << std::endl;
    std::cout << "====================================================" << std::endl;

    // Optional: load mu/sigma from .npy files at runtime
    // (overrides the compile-time constants above if files are found)
    std::vector<float> mu_runtime, sigma_runtime;
    if (load_npy_float32("pitnn_mu.npy", mu_runtime) &&
        load_npy_float32("pitnn_sigma.npy", sigma_runtime)) {
        std::cout << "[PITNN] Loaded normalisation from .npy files" << std::endl;
    } else {
        std::cout << "[PITNN] Using compile-time normalisation constants" << std::endl;
    }

    // Create controller (set use_cuda=true if CUDA is available)
    PITNNController ctrl("pitnn_scripted.pt", /*use_cuda=*/true);

    // Test operating conditions
    struct Scenario {
        float V1, V2, Pref;
        const char* label;
    };
    Scenario scenarios[] = {
        {800, 800, 10000, "10kW nominal"},
        {760, 840,  8000, "8kW V-variation"},
        {800, 800, 20000, "20kW"},
        {800, 800,  5000, "5kW light load"},
        {880, 720, 50000, "50kW high V"},
        {800, 800, 30000, "30kW mid load"},
        {840, 760, 40000, "40kW asymm V"},
        {720, 880, 15000, "15kW off-voltage"},
    };

    printf("\n%-22s %5s %5s %7s  %9s  %10s  %6s  %8s\n",
           "Condition", "V1", "V2", "Pref",
           "phi3(rad)", "delay(us)", "duty%", "t(us)");
    printf("%s\n", std::string(80, '-').c_str());

    for (const auto& s : scenarios) {
        ctrl.reset();
        float iL_est = s.V1 * s.V2 / (V1_NOM * V2_NOM) * 10.7f * 0.3f;

        // Warm up the buffer
        for (int w = 0; w < 5; w++)
            ctrl.step(s.V1, s.V2, iL_est, s.Pref);

        // Timed inference
        auto t0   = std::chrono::high_resolution_clock::now();
        auto phi  = ctrl.step(s.V1, s.V2, iL_est, s.Pref);
        auto t1   = std::chrono::high_resolution_clock::now();
        float t_us = std::chrono::duration<float, std::micro>(t1 - t0).count();

        float delay_us  = PITNNController::phi3_to_delay_us(phi[2]);
        float duty_pct  = PITNNController::phi1_to_duty_pct(phi[0]);

        printf("%-22s %5.0f %5.0f %7.0f  %9.4f  %10.3f  %5.1f%%  %8.1f\n",
               s.label, s.V1, s.V2, s.Pref,
               phi[2], delay_us, duty_pct, t_us);
    }

    std::cout << std::endl;
    std::cout << "All three angles phi1, phi2, phi3 independently predicted each cycle."
              << std::endl;
    std::cout << "Gate drive: phi3 -> phase delay  |  phi1 -> primary duty  |  phi2 -> secondary duty"
              << std::endl;

    // ── Integration example: minimal real-time loop ───────────────────────
    std::cout << "\n--- Integration Example (replace sensor stubs with real reads) ---\n";
    ctrl.reset();
    float Pref = 20000.0f;

    for (int cycle = 0; cycle < 10; cycle++) {
        // Replace these with real ADC reads
        float V1_meas = 800.0f + (cycle % 3) * 1.5f;
        float V2_meas = 798.0f + (cycle % 2) * 1.0f;
        float iL_meas = 27.5f  + (cycle % 4) * 0.5f;

        auto phi = ctrl.step(V1_meas, V2_meas, iL_meas, Pref);

        float phase_delay_s = phi[2] / (2.0f * PI_F * FSW);
        float duty          = phi[0] / PI_F;

        if (cycle % 3 == 0) {
            printf("Cycle %3d: phi3=%.4f rad  delay=%.3fus  duty=%.1f%%\n",
                   cycle, phi[2], phase_delay_s * 1e6f, duty * 100.0f);
        }

        // Apply to gate drive hardware:
        // pwm_set_primary_duty(duty);
        // pwm_set_phase_delay(phase_delay_s);
        // pwm_commit();
    }

    return 0;
}

File: pitnn_inference.py
"""
PITNN DAB Converter — Python Inference (Option 1)
==================================================
Runs the trained PITNN using only the three exported files.
No dependency on pitnn_dab.py or any training code.

Required files (all in the same folder as this script):
    pitnn_scripted.pt   — TorchScript model
    pitnn_mu.npy        — normalisation means
    pitnn_sigma.npy     — normalisation standard deviations

Install:
    pip install torch numpy

Run:
    python pitnn_inference.py

Copyright (c) 2026 Chukwuemeka Nzeadibe
Mississippi State University — All Rights Reserved
"""

import time
import numpy as np
import torch

# ─────────────────────────────────────────────────────────────────────────────
# CONSTANTS  (must match values used during training)
# ─────────────────────────────────────────────────────────────────────────────
V1_NOM    = 800.0    # V  — nominal primary bus voltage
V2_NOM    = 800.0    # V  — nominal secondary bus voltage
FSW       = 100e3    # Hz — switching frequency
PI        = 3.141592653589793
PHI12_MIN = PI * 0.65   # 2.0420 rad — lower bound for phi1/phi2
PHI12_MAX = PI * 0.99   # 3.1102 rad — upper bound for phi1/phi2
PHI_MIN   = 0.02         # rad — lower bound for phi3
PHI3_MAX  = 1.50         # rad — upper bound for phi3
SEQ_LEN   = 20           # number of past switching cycles the model sees
N_FEAT    = 8            # [V1, V2, iL, phi1, phi2, phi3, Pref, V1V2/Vnom2]


# ─────────────────────────────────────────────────────────────────────────────
# LOAD MODEL AND NORMALISATION STATS
# ─────────────────────────────────────────────────────────────────────────────

def load_pitnn(model_path="pitnn_scripted.pt",
               mu_path="pitnn_mu.npy",
               sigma_path="pitnn_sigma.npy",
               device=None):
    """
    Load the TorchScript model and normalisation constants.
    Returns (model, mu, sigma, device).
    """
    if device is None:
        device = "cuda" if torch.cuda.is_available() else "cpu"

    print(f"[PITNN] Loading model from: {model_path}")
    model = torch.jit.load(model_path, map_location=device)
    model.eval()

    mu    = np.load(mu_path).astype(np.float32)
    sigma = np.load(sigma_path).astype(np.float32)

    print(f"[PITNN] Ready on {device}")
    print(f"[PITNN] Input features: [V1, V2, iL, φ1, φ2, φ3, Pref, V1·V2/V²nom]")
    print(f"[PITNN] φ1,φ2 ∈ [{PHI12_MIN:.4f}, {PHI12_MAX:.4f}] rad  (predicted)")
    print(f"[PITNN] φ3    ∈ [{PHI_MIN:.4f},   {PHI3_MAX:.4f}]  rad  (predicted)")
    return model, mu, sigma, device


# ─────────────────────────────────────────────────────────────────────────────
# REAL-TIME CONTROLLER CLASS
# ─────────────────────────────────────────────────────────────────────────────

class PITNNInference:
    """
    Stateful real-time PITNN controller.
    Maintains a rolling 20-step history buffer.
    Call step() once per switching cycle.

    Usage:
        ctrl = PITNNInference()
        phi1, phi2, phi3 = ctrl.step(V1=800, V2=798, iL=27.5, Pref=20000)
    """

    def __init__(self, model_path="pitnn_scripted.pt",
                 mu_path="pitnn_mu.npy",
                 sigma_path="pitnn_sigma.npy",
                 device=None):
        self.model, self.mu, self.sigma, self.device = load_pitnn(
            model_path, mu_path, sigma_path, device
        )
        # Rolling buffer initialised to zeros — call prime() before first step()
        self._buffer    = np.zeros((SEQ_LEN, N_FEAT), dtype=np.float32)
        self._phi1_prev = (PHI12_MIN + PHI12_MAX) / 2.0
        self._phi2_prev = (PHI12_MIN + PHI12_MAX) / 2.0
        self._phi3_prev = 0.22

    def prime(self, V1: float, V2: float, Pref: float,
              phi1_seed: float = None, phi2_seed: float = None,
              phi3_seed: float = None):
        """
        Fill the 20-step history buffer with a plausible operating point
        before the first real step() call.

        The buffer must not start as all-zeros — the model was never trained
        on zero-padded inputs and will produce stuck or incorrect outputs if
        called cold. Prime with a physically reasonable seed instead.

        phi1_seed is derived from Pref by inverting the approximate power
        relationship: P ≈ K * (phi12/π) * phi3_peak.  Higher power → lower
        phi12, mirroring what solve_optimal_phi() returns during training.

        Parameters
        ----------
        V1, V2   : bus voltages (V)
        Pref     : power reference (W)
        phi1_seed, phi2_seed, phi3_seed : optional override angles (rad).
        """
        v_ratio = float(V1 * V2) / (V1_NOM * V2_NOM)

        if phi1_seed is None:
            # Approximate phi12 from Pref — mirrors solver: higher P → lower phi12.
            # At P_max (~70kW nominal) phi12 ≈ PHI12_MIN; at light load ≈ PHI12_MAX.
            P_max_approx = 105674.0 * v_ratio * (PHI12_MAX / PI) * 0.5 * (1 - 0.5 / 3.136)
            # Linear interpolation across the valid phi12 range
            frac = float(np.clip(Pref / max(P_max_approx, 1.0), 0.0, 1.0))
            phi1_seed = float(PHI12_MAX - frac * (PHI12_MAX - PHI12_MIN))
            phi1_seed = float(np.clip(phi1_seed, PHI12_MIN, PHI12_MAX))

        if phi2_seed is None:
            phi2_seed = phi1_seed

        if phi3_seed is None:
            # Estimate phi3 from Pref using the power model with the estimated phi12
            K_eff = 105674.0 * v_ratio * (phi1_seed / PI)
            phi3_seed = float(np.clip(Pref / max(K_eff, 1.0), PHI_MIN, PHI3_MAX))

        iL_est = float(V1 * V2 / (V1_NOM * V2_NOM) * 10.7 * phi3_seed)

        feat = np.array([V1, V2, iL_est,
                         phi1_seed, phi2_seed, phi3_seed,
                         Pref, v_ratio], dtype=np.float32)
        feat_norm = self._normalise(feat)

        # Fill entire buffer with this seed so all 20 steps look realistic
        self._buffer[:] = feat_norm
        self._phi1_prev = phi1_seed
        self._phi2_prev = phi2_seed
        self._phi3_prev = phi3_seed

    def reset(self):
        """
        Clear history buffer and prev-angle state.
        Always call prime() immediately after reset() — do NOT call step()
        on a zero buffer, the model was never trained on zero-padded inputs
        and will produce stuck outputs.
        """
        self._buffer    = np.zeros((SEQ_LEN, N_FEAT), dtype=np.float32)
        self._phi1_prev = (PHI12_MIN + PHI12_MAX) / 2.0
        self._phi2_prev = (PHI12_MIN + PHI12_MAX) / 2.0
        self._phi3_prev = 0.22

    def _normalise(self, feat: np.ndarray) -> np.ndarray:
        """Apply training normalisation: (x - mu) / sigma."""
        return ((feat - self.mu) / self.sigma).astype(np.float32)

    def step(self, V1: float, V2: float, iL: float, Pref: float,
             phi_prev: tuple = None) -> tuple:
        """
        Run one PITNN inference step.

        Parameters
        ----------
        V1        : float — primary DC bus voltage (V)
        V2        : float — secondary DC bus voltage (V)
        iL        : float — inductor current (A)
        Pref      : float — power reference from outer control loop (W)
        phi_prev  : (phi1, phi2, phi3) previous outputs (rad); uses internal
                    state if None

        Returns
        -------
        (phi1, phi2, phi3) : floats in radians — all three independently predicted
            phi1 ∈ [PHI12_MIN, PHI12_MAX]  — primary bridge inner duty
            phi2 ∈ [PHI12_MIN, PHI12_MAX]  — secondary bridge inner duty
            phi3 ∈ [PHI_MIN,   PHI3_MAX]   — external phase shift → gate drive
        """
        if phi_prev is not None:
            phi1_p, phi2_p, phi3_p = phi_prev
        else:
            phi1_p = self._phi1_prev
            phi2_p = self._phi2_prev
            phi3_p = self._phi3_prev

        # Build 8-feature state vector using previous predicted angles
        v_ratio = float(V1 * V2) / (V1_NOM * V2_NOM)
        feat = np.array([
            V1, V2, iL,
            phi1_p, phi2_p, phi3_p,
            Pref, v_ratio
        ], dtype=np.float32)

        # Shift buffer left, append new normalised state at the end
        self._buffer = np.roll(self._buffer, -1, axis=0)
        self._buffer[-1] = self._normalise(feat)

        # Single PITNN forward pass
        x = torch.from_numpy(self._buffer).unsqueeze(0).to(self.device)
        with torch.no_grad():
            phi_out = self.model(x).cpu().numpy().squeeze()

        phi1 = float(phi_out[0])
        phi2 = float(phi_out[1])
        phi3 = float(phi_out[2])
        self._phi1_prev = phi1
        self._phi2_prev = phi2
        self._phi3_prev = phi3
        return phi1, phi2, phi3

    def phi3_to_delay_us(self, phi3: float) -> float:
        """Convert φ3 (rad) to gate drive phase delay in microseconds."""
        return phi3 / (2.0 * PI * FSW) * 1e6

    def phi1_to_duty_pct(self, phi1: float) -> float:
        """Convert φ1 (rad) to inner duty cycle percentage."""
        return (phi1 / PI) * 100.0


# ─────────────────────────────────────────────────────────────────────────────
# UTILITY: PRINT NORMALISATION CONSTANTS
# ─────────────────────────────────────────────────────────────────────────────

def print_normalisation_constants(mu_path="pitnn_mu.npy",
                                  sigma_path="pitnn_sigma.npy"):
    """
    Print exact mu and sigma values for use in C++, MATLAB, or embedded code.
    Copy these numbers directly into your target-language implementation.
    """
    mu    = np.load(mu_path).astype(np.float32)
    sigma = np.load(sigma_path).astype(np.float32)

    labels = ["V1 (V)", "V2 (V)", "iL (A)", "φ1 (rad)",
              "φ2 (rad)", "φ3 (rad)", "Pref (W)", "V1V2/Vnom²"]

    print("\nNormalisation Constants (copy into C++/MATLAB/embedded code)")
    print("=" * 60)
    print(f"{'Feature':<16} {'mu':>20} {'sigma':>20}")
    print("-" * 60)
    for i, label in enumerate(labels):
        print(f"{label:<16} {mu[i]:>20.8f} {sigma[i]:>20.8f}")
    print()
    print("C++ arrays:")
    mu_str    = ", ".join(f"{v:.8f}f" for v in mu)
    sigma_str = ", ".join(f"{v:.8f}f" for v in sigma)
    print(f"float mu[8]    = {{{mu_str}}};")
    print(f"float sigma[8] = {{{sigma_str}}};")


# ─────────────────────────────────────────────────────────────────────────────
# DEMO — runs without any hardware
# ─────────────────────────────────────────────────────────────────────────────

def run_demo():
    print("=" * 66)
    print("  PITNN Inference — Python Demo (pitnn_inference.py)")
    print("=" * 66)

    # Print the normalisation constants first
    print_normalisation_constants()

    # Create controller
    ctrl = PITNNInference()

    # Test operating conditions matching the paper's inference table
    scenarios = [
        (800, 800, 10000, "10kW nominal"),
        (760, 840,  8000, "8kW voltage variation"),
        (800, 800, 20000, "20kW"),
        (800, 800,  5000, "5kW light load"),
        (880, 720, 50000, "50kW high voltage"),
        (800, 800, 30000, "30kW mid load"),
        (840, 760, 40000, "40kW asymmetric V"),
        (720, 880, 15000, "15kW off-voltage"),
    ]

    print(f"\n{'Condition':<22} {'V1':>5} {'V2':>5} {'Pref':>7}  "
          f"{'φ1 (rad)':>9}  {'φ3 (rad)':>9}  {'delay (µs)':>10}  {'duty%':>6}  {'t (µs)':>8}")
    print("-" * 92)

    for V1, V2, Pref, label in scenarios:
        ctrl.reset()

        # Prime the buffer with a physically realistic seed for this condition,
        # then run a full SEQ_LEN warmup so every buffer slot is meaningful.
        # Without this the Transformer sees mostly zeros which it was never
        # trained on, producing stuck/incorrect phi1 outputs.
        ctrl.prime(float(V1), float(V2), float(Pref))
        iL_est = float(V1 * V2 / (V1_NOM * V2_NOM) * 10.7 * ctrl._phi3_prev)

        for _ in range(SEQ_LEN):   # fill all 20 slots with real data
            ctrl.step(float(V1), float(V2), iL_est, float(Pref))

        # Timed inference
        t0 = time.perf_counter()
        phi1, phi2, phi3 = ctrl.step(float(V1), float(V2), iL_est, float(Pref))
        t_us = (time.perf_counter() - t0) * 1e6

        delay_us = ctrl.phi3_to_delay_us(phi3)
        duty_pct = ctrl.phi1_to_duty_pct(phi1)

        print(f"{label:<22} {V1:>5.0f} {V2:>5.0f} {Pref:>7.0f}  "
              f"{phi1:>9.4f}  {phi3:>9.4f}  {delay_us:>10.3f}  {duty_pct:>5.1f}%  {t_us:>8.1f}")

    print(f"\nAll three angles φ1, φ2, φ3 independently predicted each cycle.")
    print(f"Gate drive: φ3 → phase delay  |  φ1 → primary duty  |  φ2 → secondary duty")


# ─────────────────────────────────────────────────────────────────────────────
# INTEGRATION EXAMPLE — drop this into your own control loop
# ─────────────────────────────────────────────────────────────────────────────

def integration_example():
    """
    Minimal integration example.
    Replace the simulated sensor values with your real ADC readings.
    """
    ctrl = PITNNInference()

    # Your real-time loop — call at fsw = 100kHz
    Pref = 20000.0   # W — from your outer PI voltage controller

    # Prime the buffer before the first step — critical for correct predictions
    ctrl.prime(V1=800.0, V2=800.0, Pref=Pref)

    for cycle in range(20):
        # ── READ SENSORS (replace with real ADC calls) ────────────────────
        V1_meas = 800.0 + (cycle % 3) * 1.5     # simulated noise
        V2_meas = 798.0 + (cycle % 2) * 1.0
        iL_meas = 27.5  + (cycle % 4) * 0.5

        # Replace with actual ADC values
        # V1_meas = adc.read(channel=0) * V1_SCALE
        # V2_meas = adc.read(channel=1) * V2_SCALE
        # iL_meas = adc.read(channel=2) * IL_SCALE

        # ── PITNN INFERENCE ───────────────────────────────────────────────
        phi1, phi2, phi3 = ctrl.step(V1_meas, V2_meas, iL_meas, Pref)

        # ── CONVERT TO GATE DRIVE TIMING ─────────────────────────────────
        phase_delay_s = phi3 / (2.0 * PI * FSW)      # seconds
        inner_duty    = phi1 / PI                     # fraction (0 to 1)

        # ── APPLY TO PWM HARDWARE (replace with your hardware API) ────────
        # pwm.set_primary_duty(inner_duty)
        # pwm.set_phase_delay(phase_delay_s)
        # pwm.commit()

        if cycle % 5 == 0:
            print(f"Cycle {cycle:3d}: φ3={phi3:.4f} rad  "
                  f"delay={phase_delay_s*1e6:.3f}µs  duty={inner_duty*100:.1f}%")


if __name__ == "__main__":
    run_demo()
    print("\n" + "=" * 66)
    print("  Integration Example (replace sensor stubs with real ADC reads)")
    print("=" * 66)
    integration_example()
